Parameter for keep waypoints on the left/right

fijemax opened the pull request:
# Description : 
This PR adds a parameter for each point to chose the start/stop side.
Three option can be used : 
  - DEFAULT to start/stop at the drive side
  - OPPOSITE to start/stop at the opposite of default (cross the road)
  - BOTH each side can be used (old behavior).

The default depending at the country.

# Issue
This pull request is linking to this issue : 
https://github.com/Project-OSRM/osrm-backend/issues/2201

## Tasklist
 - [x] Adding sides parameter into base parameters, it can take the values DEFAULT, OPPOSITE or BOTH.
 - [x] Adding url parser for "sides" parameter, url values are "d" for DEFAULT, "o" for OPPOSITE and "b" for BOTH, example : "sides=s;s".
 - [x] Checking parameters, if "sides" parameter is used, the number of parameter is the same as number of location.
 - [x] Create a phantom to start/stop at side drive or opposite drive side.
 - [x] Retrieve the country's drive side for apply the DEFAULT.

**For the moment the country's drive side is set to right by default, we can find this parameter in the lua file.
How can access this parameter in the osrm-routed ?**

Example for DEFAULT on start and stop route point value : 
Here start and stop 
![capture d ecran de 2017-05-12 16-07-54](https://cloud.githubusercontent.com/assets/16467651/26003333/b455333e-3732-11e7-93f9-e4b4b854104d.png)
![capture d ecran de 2017-05-12 16-07-32](https://cloud.githubusercontent.com/assets/16467651/26003399/d85d748a-3732-11e7-8ad8-4338fec88b12.png)

Another example : 
![capture d ecran de 2017-05-12 18-00-44](https://cloud.githubusercontent.com/assets/16467651/26006668/9c129766-373d-11e7-9132-481bdd81335d.png) ![capture d ecran de 2017-05-12 18-00-51](https://cloud.githubusercontent.com/assets/16467651/26006669/9c17b804-373d-11e7-8898-345199e4c0de.png)



**For the one-way roads :** 
The algorythm do not use the "side" parameter for the phantom node option because the driver hasn't the choice.
![capture d ecran de 2017-05-12 16-11-29](https://cloud.githubusercontent.com/assets/16467651/26003514/34c6fde0-3733-11e7-8234-1fcf5317a8e4.png)
![capture d ecran de 2017-05-12 16-11-37](https://cloud.githubusercontent.com/assets/16467651/26003515/34c91f3a-3733-11e7-9996-c6437eab71d7.png)

fijemax commented:
@frodrigo

TheMarex commented:
Thank you for taking a stab at this! 🎉 I just skimmed the diff quickly some things that I noticed:

1. No test coverage, this should have a cucumber test
2. No URL parser test, this should have a unit-test
3. NodeJS bindings are not extended yet.

I can put a real review on my TODO for later today, I need to think a little bit about parameter naming and such.

fijemax commented:
Hi @TheMarex 
I I extended the NodeJS  for the sides parameter.
I added the cucumber test.

oxidase commented on features/testbot/side_param.feature:
```diff
@@ -0,0 +1,207 @@
+
+        When I route I should get
+            | from | to | sides | route    |
+            | s    | e  | b o   | ab,bc,bc |
```
is it possible to add a turn instructions check? to see explicitly a u-turn

oxidase commented on include/engine/geospatial_query.hpp:
```diff
@@ -208,16 +211,50 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
+                    //                       |
+                    //                 input_coordinate
+
+                    bool input_coordinate_is_at_right = !util::coordinate_calculation::isCCW(
```
`util::coordinate_calculation::isCCW(coordinates[segment.data.u], input_coordinate, coordinates[segment.data.v]);` to correspond the figure above

fijemax commented:
I just added the nodeJS binding.

fijemax commented on include/server/api/base_parameters_grammar.hpp:
```diff
@@ -149,10 +159,13 @@ struct BaseParametersGrammar : boost::spirit::qi::grammar<Iterator, Signature>
             qi::lit("bearings=") >
             (-(qi::short_ > ',' > qi::short_))[ph::bind(add_bearing, qi::_r1, qi::_1)] % ';';
 
+        sides_rule = qi::lit("sides=") >
```
I tryed to change by this code :
```
        side_rule = qi::char_("a-zA-Z")[qi::_val = engine::Side{engine::BOTH}];
        struct StartSideValues: qi::symbols<char, engine::Side>
        {
            StartSideValues() {add("b", engine::Side{engine::BOTH})("d", engine::Side{engine::DEFAULT})("o", engine::Side{engine::OPPOSITE})("", engine::Side{engine::OPPOSITE}); }
        };
        StartSideValues values;
        sides_rule = qi::lit("sides=") >
                        (-(values | side_rule) %
                         ';')[ph::bind(&engine::api::BaseParameters::sides, qi::_r1) = qi::_1];
```
But it provokes segfault.
I am trying to analyse with gdb but it's a really complicated backtrace.

oxidase replied:
@fijemax it crashes because grammars are local variables in the constructor.

What i suggest is to use in the way as in commit https://github.com/oxidase/osrm-backend/commit/301a1cca6c5a6f3ee229213448577d9753e337d5 It contains more restrictive grammar [bod|eps]

Please add also some unit tests for parsing incorrect URLs as suggested by @TheMarex

fijemax commented:
Hi @oxidase, thanks for parser example.
I refactored the code 
 - Simple enumerator for Side
 - Change gi::lit to qi::symbol for side parser
I added URL tests in the unit_tests

oxidase commented on unit_tests/server/parameters_parser.cpp:
```diff
@@ -46,7 +46,7 @@ BOOST_AUTO_TEST_CASE(invalid_route_urls)
                       32UL);
     BOOST_CHECK_EQUAL(testInvalidOptions<RouteParameters>("1,2;3,4?overview=false&sides=foo"),
-                      30UL);
+                      29UL);
```
:+1:

oxidase commented on include/engine/side.hpp:
```diff
@@ -0,0 +1,47 @@
+#ifndef OSRM_ENGINE_SIDE_HPP
+#define OSRM_ENGINE_SIDE_HPP
+
+#include <string>
```
It is not needed

oxidase commented on include/server/api/base_parameters_grammar.hpp:
```diff
@@ -149,10 +149,17 @@ struct BaseParametersGrammar : boost::spirit::qi::grammar<Iterator, Signature>
+        sides_rule = qi::lit("sides=") >
+                        (-(side_type | qi::attr(engine::Side::BOTH)) %
+                         ';')[ph::bind(&engine::api::BaseParameters::sides, qi::_r1) = qi::_1];
+                     //(-qi::symbols[qi::char_("aA-zZ")])[ph::bind(add_side, qi::_r1, qi::_1)] % ';';
```
not needed

oxidase commented on include/server/api/base_parameters_grammar.hpp:
```diff
@@ -149,10 +149,17 @@ struct BaseParametersGrammar : boost::spirit::qi::grammar<Iterator, Signature>
 
+        side_type.add("d", engine::Side::DEFAULT)("o", engine::Side::OPPOSITE)("b", engine::Side::BOTH);
+        sides_rule = qi::lit("sides=") >
+                        (-(side_type | qi::attr(engine::Side::BOTH)) %
```
Here must be either `-` for `vector<optional<Side>>` or `qi::attr(engine::Side::BOTH)`  for `vector<Side>` 
`-` will place empty optionals for empty values, but `attr(BOTH)` will place `BOTH`.

`(-side_type %';')[...]` is enough for vector of optionals

TheMarex commented:
Thanks for adding the tests! Some API features need a quick discussion:

1. Is there a real use-case for `opposite`? I feel the actual feature here is the ability to force the curb approach.
2. Parameter name. I feel something like `approach={curb,unrestriced}` might be more descriptive.
3. Do we need the ability to configure this on per-waypoint basis? The scenarios I could come up with (taxi, garbage can pickup) would only need this to be a global flag that applies to all waypoints.

frodrigo commented:
> 1. Is there a real use-case for opposite? I feel the actual feature here is the ability to force the curb approach.

Yes we build it for a use case where we need the curb approach. We do not have use case the opposite. But why implement half an option as we can easily make it for both sides? Nevertheless we can remove opposite if you want.

> 3. Do we need the ability to configure this on per-waypoint basis? The scenarios I could come up with (taxi, garbage can pickup) would only need this to be a global flag that applies to all waypoints.

We need it per waypoint. E.g. on garbage pickup, the pickup point must be done with the curb approach, but at the depot the side does not matter.

We also plan to add to this PR the support of the approach to other plugins. Especially on the table plugin (approach per waypoint make even more sens on it).

TheMarex commented:
> But why implement half an option as we can easily make it for both sides?

One of the main concerns with adding stuff is usually API-surface area. The fewer options we have, the less likely it is that we need to do a breaking API change. Would be cool if we could limit this to only `unrestricted` and `curb`.

> We need it per waypoint. E.g. on garbage pickup, the pickup point must be done with the curb approach, but at the depot the side does not matter.

Okay thanks for the explanation. /cc @daniel-j-h because he has been doing a lot of VRP stuff lately and this might be useful for him too?

> We also plan to add to this PR the support of the approach to other plugins.

🤔 I think that should already work if the right method was extended, all plugins rely on `GetPhantomNodes()` which should operate on the `BaseParameter` (hence works for all plugins, similar to how we handle `bearings` and `radiuses`)

---------------------------

I would like to propose the following API:

### HTTP

- `approaches={approach}[;{approach}; ...]` where `approach` is a `String` of either `curb` or `unrestricted` (default: `unrestricted`). Needs to have the same length as the number of coordinates given.

Example: `?approaches=curb;unrestricted;curb;curb;curb;unrestricted` or`?approaches=curb;;curb;curb;curb;`

### Node JS

- `approaches` **Array** of **String**s that can either be `curb` or `unrestricted`. Needs to have the same length as the number of coordinates given. (default: `null`, which means `unrestricted`)

Example:  `{approaches: ["curb", "unrestricted", "curb", "curb"]}` or `{approaches: ["curb", null, "curb", "curb"]}`

fijemax commented:
Hi @TheMarex ,
Thank you for the review, I going to make the changes as respecting your API descriptions.
Also, I will don't know how to retrieve the country's drive side for apply the `curb` option, this is actualy possible ?
For the moment I apply right side by default for `curb` approach.

TheMarex commented:
> Also, I will don't know how to retrieve the country's drive side for apply the curb option, is this actually possible?

Ah this is part of `ProfileProperties::left_hand_driving`. To use this you would need to extend `ContiguousInternalMemoryDataFacadeBase` and `BaseDataFacade` with a new function like `bool IsLeftHandDriving() { return m_profile_properties.left_hand_driving; }`.

It can be configured in the profile via `properties.left_hand_driving = true`.

daniel-j-h commented:
> We need it per waypoint. E.g. on garbage pickup, the pickup point must be done with the curb approach, but at the depot the side does not matter.
> 
> Okay thanks for the explanation. /cc @daniel-j-h because he has been doing a lot of VRP stuff lately and this might be useful for him too?

Yeah I can totally see how this is useful. You can somewhat fake it already setting bearings - but then you need to do some pre-processing / set the bearing constraints manually.

emiltin commented:
note that ProfileProperties::left_hand_driving is a global property set once by the profile. it doesn't take into account what country the current request (and and waypoint) is located in. so if you're working with a g planet dataset, the api will be wrong in some countries, and right in others.

fijemax commented:
@oxidase 
I refactored the API and propagate the param for every phantom nodes search function.
The support for left hand driving is ok.
Cucumber tests are ok and url unit tests too.

TheMarex commented:
@fijemax looks like the only thing that is failing right now is the formating check. Can you try:

```
./scripts/format.sh
```

That requires `clang-format` 3.8.

TheMarex commented on include/engine/approach.hpp:
```diff
@@ -0,0 +1,44 @@
+namespace engine
+{
+
+enum Approach
```
`enum class Approach : std::uint8_t` would be better here, for one to remove implicit conversions and to limit the storage requirements to one byte per coordinate.

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
                                const int bearing,
-                               const int bearing_range) const override final
+                               const int bearing_range,
+                               const engine::Approach approach) const override final
```
`engine::` is unnecessary.

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
     NearestPhantomNodes(const util::Coordinate input_coordinate,
-                        const unsigned max_results) const override final
+                        const unsigned max_results,
+                        const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
                         const unsigned max_results,
-                        const double max_distance) const override final
+                        const double max_distance,
+                        const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
                         const int bearing,
-                        const int bearing_range) const override final
+                        const int bearing_range,
+                        const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
                         const int bearing,
-                        const int bearing_range) const override final
+                        const int bearing_range,
+                        const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
-        const util::Coordinate input_coordinate) const override final
+    std::pair<PhantomNode, PhantomNode>
+    NearestPhantomNodeWithAlternativeFromBigComponent(const util::Coordinate input_coordinate,
+                                                      const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
 
     std::pair<PhantomNode, PhantomNode> NearestPhantomNodeWithAlternativeFromBigComponent(
-        const util::Coordinate input_coordinate, const double max_distance) const override final
+        const util::Coordinate input_coordinate, const double max_distance, const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -650,108 +651,117 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
                                                       const int bearing,
-                                                      const int bearing_range) const override final
+                                                      const int bearing_range,
+                                                      const engine::Approach approach) const override final
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
     NearestPhantomNodesInRange(const util::Coordinate input_coordinate,
-                               const float max_distance) const = 0;
+                               const float max_distance,
+                               const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
                         const int bearing,
-                        const int bearing_range) const = 0;
+                        const int bearing_range,
+                        const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
                         const int bearing,
-                        const int bearing_range) const = 0;
+                        const int bearing_range,
+                        const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
     NearestPhantomNodes(const util::Coordinate input_coordinate,
-                        const unsigned max_results) const = 0;
+                        const unsigned max_results,
+                        const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
                         const unsigned max_results,
-                        const double max_distance) const = 0;
+                        const double max_distance,
+                        const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
+    virtual std::pair<PhantomNode, PhantomNode>
+    NearestPhantomNodeWithAlternativeFromBigComponent(const util::Coordinate input_coordinate,
+                                                      const double max_distance,
+                                                      const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
                                                       const int bearing,
-                                                      const int bearing_range) const = 0;
+                                                      const int bearing_range,
+                                                      const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/datafacade/datafacade_base.hpp:
```diff
@@ -93,44 +94,54 @@ class BaseDataFacade
                                                       const int bearing,
-                                                      const int bearing_range) const = 0;
+                                                      const int bearing_range,
+                                                      const engine::Approach approach) const = 0;
```
`engine::`

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -50,14 +53,17 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
     NearestPhantomNodesInRange(const util::Coordinate input_coordinate,
-                               const double max_distance) const
+                               const double max_distance,
+                               const engine::Approach approach) const
```
`engine::`

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -69,13 +75,17 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
                                const int bearing,
-                               const int bearing_range) const
+                               const int bearing_range,
+                               const engine::Approach approach) const
```
`engine::`

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -91,13 +101,15 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
                         const int bearing,
-                        const int bearing_range) const
+                        const int bearing_range,
+                        const engine::Approach approach) const
```
`engine::`

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -134,11 +148,15 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
-    NearestPhantomNodes(const util::Coordinate input_coordinate, const unsigned max_results) const
+    NearestPhantomNodes(const util::Coordinate input_coordinate,
+                        const unsigned max_results,
+                        const engine::Approach approach) const
```
`engine::`

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -151,11 +169,14 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
                         const unsigned max_results,
-                        const double max_distance) const
+                        const double max_distance,
+                        const engine::Approach approach) const
```
`engine::`

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -15,6 +16,8 @@
 #include <memory>
 #include <vector>
 
+#include <util/log.hpp>
```
not needed anymore?

TheMarex commented on include/engine/geospatial_query.hpp:
```diff
@@ -545,6 +578,32 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
         return datafacade.GetComponentID(data.forward_segment_id.id).is_tiny;
     }
 
+    std::pair<bool, bool> approach_usage(const util::Coordinate &input_coordinate,
```
Rename to `ApproachUsage` for camel-case. A better name could be `CheckApproach`.

TheMarex commented on include/nodejs/node_osrm_support.hpp:
```diff
@@ -297,6 +298,63 @@ inline bool argumentsToParameter(const Nan::FunctionCallbackInfo<v8::Value> &arg
+
+            if (approach_raw->IsNull())
+            {
+                params->bearings.emplace_back();
```
Copy&Paste error, should be `params->approaches`.

TheMarex commented on include/nodejs/node_osrm_support.hpp:
```diff
@@ -297,6 +298,63 @@ inline bool argumentsToParameter(const Nan::FunctionCallbackInfo<v8::Value> &arg
+
+        if (!approaches->IsArray())
+        {
+            Nan::ThrowError("Approaches must be an array of arrays of numbers");
```
Error message is wrong: Should be `arrays of strings`

TheMarex commented on include/nodejs/node_osrm_support.hpp:
```diff
@@ -297,6 +298,63 @@ inline bool argumentsToParameter(const Nan::FunctionCallbackInfo<v8::Value> &arg
+                }
+                else
+                {
+                    Nan::ThrowError("'approach' param must be one of [curb, unrestricted]");
```
`approach` -> `approaches`

fijemax commented:
Hi @TheMarex 
Change was done.
I don't konw how testing the nodejs binding ?

TheMarex commented:
@fijemax for some examples look at `test/nodejs/route.js`. What you basically want is to create a request that uses your parameter and try a few permutations that would create an error.

fijemax commented:
@TheMarex  When I run nodejs `test/nodes/index.js` or `nodejs test/nodes/route.js` I have the following error : 
`Error: Cannot find module './binding/node_osrm.node'`

fijemax commented:
It's ok, I hadn't activate ENABLE_NODE_BINDINGS during cmake compilation.

TheMarex commented:
@fijemax awesome, this looks good to merge for me. 🎉

daniel-j-h commented:
Very cool - looking forward to having this in a stable release!

What I just now realized after this was merged is we need
- [ ] a changelog entry (changelog.md)
- [ ] docs for the http api (docs/http.md)
- [ ] docs for the node bindings (docs/nodejs/api.md)

could you run this out or ticket and tag it for the next release @fijemax so we don't forget about it.

(also feel free to write a osm diary / blog post about this if you have the time for it, I can totally see people being interested in this feature :) I wrote a bit [about bearings](https://www.openstreetmap.org/user/daniel-j-h/diary) - this one would fit right in)

fijemax commented:
@daniel-j-h I just created this PR #4104.

fijemax commented:
Hi @daniel-j-h, i wrote [this](https://www.openstreetmap.org/user/Jean-Maxime%20FILLAU/diary/41741) about the new feature.

daniel-j-h commented:
Sweet, thank you!

asaveljevs commented:
> 1. Is there a real use-case for opposite? I feel the actual feature here is the ability to force the curb approach.

We have tried to describe our use case for opposite in #5616, which is needed for solving a VRP problem.

geocept commented:
Is there a reason why this great feature is not in the stable release of OSRM?? I mean 6,5 years have passed now and many many people are waiting for it. Would be great if someone would bring it to stable release.

frodrigo commented:
> Is there a reason why this great feature is not in the stable release of OSRM?? I mean 6,5 years have passed now and many many people are waiting for it. Would be great if someone would bring it to stable release.

It was released in 5.8.0
https://github.com/Project-OSRM/osrm-backend/blob/master/CHANGELOG.md#580
